New architecture

jacquerie opened the pull request:
Following #2, here's my proposal for `invenio-matcher`.

What's missing:
- [x] A good commit message.
- [x] Some docstrings.
- [x] Proper error reporting (the error messages are just `TODO`).
- [x] Making the tests pass on Travis
- [x] Packaging ~~and distribution~~ (I might need @jalavik's assistance for this).

Also, since I ran cookiecutter a while ago some boilerplate is not up to date. I opened #5 to track this.

jacquerie committed:
global: initial architecture

* Implements the architecture proposed in issue #2: an API with
  a single `match` method that accepts keyword arguments and
  dispatches them to a configurable engine. (closes #2)

* Removes all default configuration. (closes #1)

Signed-off-by: Jacopo Notarstefano <jacopo.notarstefano@cern.ch>

jalavik commented on invenio_matcher/api.py:
```diff
@@ -24,22 +24,50 @@
-from .config import MATCHER_QUERIES
-from .core import execute
-from .errors import NoQueryDefined
+from invenio_matcher.core import execute
```
This can be one import statement like:

``` python
from invenio_matcher.core import execute, get_queries
```

jalavik replied:
Also, it can be a relative import even:

``` python
from .core import execute, get_queries
```

jalavik commented on invenio_matcher/api.py:
```diff
@@ -24,22 +24,50 @@
+        queries = get_queries(**kwargs)
+
+        if not queries:
             raise NoQueryDefined('TODO')
```
Here we can add the one-liner message or just raise without message.

jalavik commented on invenio_matcher/core.py:
```diff
@@ -24,78 +24,57 @@
 
-from .errors import InvalidQuery
-from .utils import build_exact_query, build_fuzzy_query
+from invenio_matcher.engine import exact
```
Combine all these imports into one statement, and make it relative ideally. E.g.

``` python
from .engine import (
    one,
    two,
    ...
)
```

jacquerie replied:
RE imports: yes, I tried to do something weird with the imports to see what they looked like. Turns out, pretty bad. Will revert to something more sane.

jalavik commented on invenio_matcher/matcherext/engines/es.py:
```diff
@@ -0,0 +1,203 @@
+
+def queries(**kwargs):
+    """Return queries defined for the given index and doc_type."""
+    try:
```
As mentioned IRL, if it's a required argument you can just add it to the method signature.

jacquerie replied:
:+1:

jacquerie replied:
Wait. `engine.py` can not know that it should call `exact(index, doc_type, **kwargs)`, since it's an implementation detail specific to Elasticsearch, so it has to call `exact(**kwargs)`. What I can do is move the logic that checks for required keyword arguments in a decorator, so that the code reads nicer. Would that be ok?

jalavik replied:
Actually, it should still work with `exact(index, doc_type, **kwargs)` because the caller does `exact(**kwargs)` and Python will align the keyword arguments accordingly.

jacquerie replied:
Yeah, I realized it one minute after writing that. Sorry! Changing all signatures. Tests are making it a breeze.

jalavik commented on invenio_matcher/matcherext/engines/es.py:
```diff
@@ -0,0 +1,203 @@
+
+from flask import current_app
+
+from invenio.base.wrappers import lazy_import
```
Invenio imports like `invenio.base` and friends needs to be ported to their counterparts `invenio_base` (with underscore).

jacquerie replied:
Oh, right. Inertia.

jacquerie commented:
I addressed the comments about the imports and the function signatures. Everything passes locally, let's see what Travis thinks (I expect to fail, I probably have to install some packages).

jacquerie commented:
Sigh, almost there. I'm having trouble mocking Elasticsearch on Travis. See you on friday!

jalavik commented on .travis.yml:
```diff
@@ -43,25 +46,27 @@ python:
 script:
-  - "sphinx-build -qnN docs docs/_build/html"
+  # FIXME(jacquerie): update to latest cookiecutter boilerplate.
+  # - "sphinx-build -qnN docs docs/_build/html"
```
Run `check-manifest` locally (to add initial `MANIFEST.in`) and here in the travis check.

jacquerie replied:
:+1:

jalavik commented on setup.py:
```diff
@@ -36,15 +36,18 @@
     'Flask>=0.10.1',
     'six>=1.7.2',
-    'Invenio>=2.0.3',
+    'invenio-base>=0.3.1',
```
`invenio-base>=0.3.1,<1.0.0` everywhere. (check syntax)

jacquerie replied:
:+1:

jacquerie commented:
Ready? @jalavik

jalavik commented on .travis.yml:
```diff
@@ -43,26 +46,38 @@ python:
 after_success:
   - coveralls
+
+deploy:
```
I think this part won't work unless the hash is generated from the repository locally, or at least until registered on PyPi manually. So you can remove it for now. It can be added again upon initial release (which will happen soon after this PR).

jacquerie replied:
Oh, didn't know that. I just C-c C-v'd it from `invenio-grobid`.

jacquerie replied:
:+1:
